types/enigma.js/index.d.ts

```diff
@@ -0,0 +1,196 @@
+         */
+        extend?: [any];
+
+        /**
+         * mixin.override is an object containing methods that overrides existing API methods.
+         * These method names needs to exist already* or engima.js will throw an error.
+         * Be careful when overriding, you may break expected behaviors in other mixins or your application.
+         * base is a reference to the previous mixin method, can be used to invoke the mixin chain before this mixin method.
+         */
+        override?: [any];
+    }
+
+    interface IProtocol {
+        // Set to false to disable the use of the bandwidth-reducing delta protocol.
+        delta?: boolean;
+    }
+
+    /**
+     * This section describes the configuration object that is sent into enigma.create(config).
+     */
+    interface IConfig {
+        /**
+         * Object containing the specification for the API to generate. Corresponds to a specific version of the QIX Engine API.
+         */
+        schema: object;
+        /**
+         * String containing a proper websocket URL to QIX Engine.
+         */
+        url: string;
+        /**
+         * A function to use when instantiating the WebSocket, mandatory for Node.js.
+         */
+        createSocket?: any;
+        /**
+         * ES6-compatible Promise library.
+         */
+        Promise?: any;
+        /**
+         * Set to true if the session should be suspended instead of closed when the websocket is closed.
+         */
+        suspendOnClose?: boolean;
+        /**
+         * Mixins to extend/augment the QIX Engine API.
+         * See Mixins section for more information how each entry in this array should look like.
+         * Mixins are applied in the array order.
+         */
+        mixins?: [any];
+        /**
+         * Interceptors for augmenting responses before they are passed into mixins and end-users.
+         * See Interceptors section for more information how each entry in this array should look like.
+         * Interceptors are applied in the array order.
+         */
+        interceptors?: [any];
+        /**
+         * An object containing additional JSON-RPC request parameters.
+         * protocol.delta :  Set to false to disable the use of the bandwidth-reducing delta protocol.
+         */
+        protocol?: any;
+    }
+
+    interface ISession {
+        /**
+         * Establishes the websocket against the configured URL. Eventually resolved with the QIX global interface when the connection has been established.
+         * @return Promise.
+         */
+        open(): Promise<IGeneratedAPI>;
+
+        /**
+         * Establishes the websocket against the configured URL. Eventually resolved with the QIX global interface when the connection has been established.
+         * @return Promise.
+         */
+        open<T extends IGeneratedAPI>(): Promise<T>;
```

Comment:
To be pedantic, this probably doesn't need to be generic; a cast to `Promise<WhateverTypeYouNeed>` could achieve the same thing.

Reply:
@DanielRosenwasser thx, this is not pedantic. I do that because the most users will use that together with @types/qlik-engineapi and this exports for example IGlobal that extends IGenericAPI, so that if they use the gernic open, they don't have to "as" the result.

Reply:
I think @DanielRosenwasser's point is that the user has to explicitly write either `open<IGlobal>()` or `open() as Promise<IGlobal>`. There is no way for the compiler to infer the desired type, so both are equivalent to a cast. That means the first is preferable to reduce confusion among people who are used to C# (or any system with non-erased generics) where `open<IGlobal>()` *is* safer than the cast.

This is quite a minor point, though, so there's no reason to hold the merge for it.

Reply:
@sandersn Thanks, now I got the point and removed the open() declaration.